Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,308 @@
+
+- `veblenWith f 0 = f`.
+- `veblenWith f a` enumerates the fixed points of `veblenWith f b` for `b < a` when `a ≠ 0`.
+-/
+@[pp_nodot]
+def veblenWith (f : Ordinal.{u} → Ordinal.{u}) (o : Ordinal.{u}) : Ordinal.{u} → Ordinal.{u} :=
+  if o = 0 then f else derivFamily fun x : Set.Iio o ↦ veblenWith f x.1
+termination_by o
+decreasing_by exact x.2
+
+@[simp]
+theorem veblenWith_zero (f : Ordinal → Ordinal) : veblenWith f 0 = f := by
+  rw [veblenWith, if_pos rfl]
+
+theorem veblenWith_of_ne_zero (f : Ordinal → Ordinal) (h : o ≠ 0) :
+    veblenWith f o = derivFamily fun x : Set.Iio o ↦ veblenWith f x.1 := by
+  rw [veblenWith, if_neg h]
+
+/-- `veblenWith f o` is always normal for `o ≠ 0`. -/
+theorem isNormal_veblenWith' (f : Ordinal → Ordinal) (h : o ≠ 0) : IsNormal (veblenWith f o) := by
+  rw [veblenWith_of_ne_zero f h]
+  exact isNormal_derivFamily _
+
+/-- `veblenWith f o` is always normal whenever `f` is. -/
+theorem isNormal_veblenWith (hf : IsNormal f) (o : Ordinal) : IsNormal (veblenWith f o) := by
+  obtain rfl | h := eq_or_ne o 0
+  · rwa [veblenWith_zero]
+  · exact isNormal_veblenWith' f h
+
+theorem veblenWith_veblenWith_of_lt (hf : IsNormal f) (h : a < b) (c : Ordinal) :
+    veblenWith f a (veblenWith f b c) = veblenWith f b c := by
+  let x : {a // a < b} := ⟨a, h⟩
+  rw [veblenWith_of_ne_zero f h.bot_lt.ne',
+    derivFamily_fp (f := fun y : Set.Iio b ↦ veblenWith f y.1) (i := x)]
+  exact isNormal_veblenWith hf x
+
+theorem veblenWith_succ (hf : IsNormal f) (o : Ordinal) :
+    veblenWith f (Order.succ o) = deriv (veblenWith f o) := by
+  rw [deriv_eq_enumOrd (isNormal_veblenWith hf o), veblenWith_of_ne_zero f (succ_ne_zero _),
+    derivFamily_eq_enumOrd]
+  · apply congr_arg
+    ext a
+    rw [Set.mem_iInter]
+    use fun ha ↦ ha ⟨o, Order.lt_succ o⟩
+    rintro (ha : _ = _) ⟨b, hb : b < _⟩
+    obtain rfl | hb := Order.lt_succ_iff_eq_or_lt.1 hb
+    · rw [Function.mem_fixedPoints_iff, ha]
+    · rw [← ha]
+      exact veblenWith_veblenWith_of_lt hf hb _
+  · exact fun a ↦ isNormal_veblenWith hf a
+
+theorem veblenWith_right_strictMono (hf : IsNormal f) (o : Ordinal) :
+    StrictMono (veblenWith f o) := by
+  obtain rfl | h := eq_or_ne o 0
+  · rw [veblenWith_zero]
+    exact hf.strictMono
+  · rw [veblenWith_of_ne_zero f h]
+    exact derivFamily_strictMono fun a ↦ isNormal_veblenWith hf a
+
+theorem veblenWith_right_monotone (hf : IsNormal f) (o : Ordinal) :
+    Monotone (veblenWith f o) :=
+  (veblenWith_right_strictMono hf o).monotone
+
+theorem veblenWith_lt_veblenWith_right_iff (hf : IsNormal f) :
+    veblenWith f o a < veblenWith f o b ↔ a < b :=
+  (veblenWith_right_strictMono hf o).lt_iff_lt
+
+theorem veblenWith_le_veblenWith_right_iff (hf : IsNormal f) :
+    veblenWith f o a ≤ veblenWith f o b ↔ a ≤ b :=
+  (veblenWith_right_strictMono hf o).le_iff_le
+
+theorem veblenWith_inj (hf : IsNormal f) :
+    veblenWith f o a = veblenWith f o b ↔ a = b :=
+  (veblenWith_right_strictMono hf o).injective.eq_iff
```

Comment:
Could you add an `injective` version of this, with conclusion `Function.Injective (veblenWith f o)`?